plotter/heat.go

```diff
@@ -0,0 +1,210 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"image/color"
+	"math"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/palette"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// GridXYZ describes three dimensional data where the X and Y
+// coordinates are arranged on a rectangular grid.
+type GridXYZ interface {
+	// Dims returns the dimensions of the grid.
+	Dims() (c, r int)
```

Comment:
Thanks for making me think more about the loop excision algorithm. What I have is incorrect, but not for that reason.

There are three cases for a double cross over (x and y mark cross over points):

1-2x-3-4-5y-6-7-8y-9-10x-11
(expect 1-2-10-11, 2-3-4-5-8-9-10-2 and 5-6-7-8-5)

1-2xy-3-4-5-6-7-8y-9-10x-11
(expect 1-2-10-11, 2-3-4-5-6-7-8-2 and 8-9-10-8)

1-2y-3-4-5x-6-7-8y-9-10x-11
(1-2-8-9-10-11, 2-3-4-5-6-7-8-2 and 5-6-7-8-9-10-5, but will get 1-2-8-9-10-11, 2-3-4-5-6-7-8-2 and 10 which is not valid since has only one point).

The current code handles the first two cases correctly, and case 2 would be broken by removing elements from seen. I suspect the last case is rare, but I can't see why it should not happen in theory.

I'm not entirely sure how to handle case three. Obviously the working list can't be altered in these cases or you lose the capacity to make the second closing loop. It looks like a sequence reconstruction problem, so people at work might have ideas.

I will add some tests for these cases and others.